### DISCUSSION TITLE
Cannot process pdf with a longer header

### DISCUSSION THREAD

**Discussion by andrecolza** [spatie/browsershot#868] [Bugs]
When the headerTemplate passed as option to Browsershot is too long (this happens for example when the base64 encoding of an image is on average long or by adding multiple images with a short encoding that individually might work) the following error occurs (omitted the full content of headerTemplate to make the error readable):

> The command "node "C:\laragon\www\hts\vendor\spatie\browsershot\src/../bin/browser.cjs" "{""url"":""file:\/\/C:\\Users\\wanco\\AppData\\Local\\Temp\\1890731036-0139898001723482615\\index.html"",""action"":""pdf"",""options"":{""args"":[],""viewport"":{""width"":800,""height"":600},""displayHeaderFooter"":true,""headerTemplate"":"...",""margin"":{""top"":""50mm"",""right"":""10mm"",""bottom"":""12mm"",""left"":""10mm""},""printBackground"":true}}"" failed. Exit Code: 1(General error) Working directory: C:\laragon\www\hts\public Output: ================ Error Output: ================ C:\laragon\www\hts\vendor\spatie\browsershot\bin\browser.cjs:13 const request = args[0].startsWith('-f ') ^ TypeError: Cannot read properties of undefined (reading 'startsWith') at Object.<anonymous> (C:\laragon\www\hts\vendor\spatie\browsershot\bin\browser.cjs:13:25) at Module._compile (node:internal/modules/cjs/loader:1369:14) at Module._extensions..js (node:internal/modules/cjs/loader:1427:10) at Module.load (node:internal/modules/cjs/loader:1206:32) at Module._load (node:internal/modules/cjs/loader:1022:12) at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) at node:internal/main/run_main_module:28:49 Node.js v20.12.2 

This prevents me from creating elaborate or multi-image headers or footers, can anyone help me on this?

**Comment by bluesheep100** [comment]
I tried recreating this the following way:
```php
$targetPath = __DIR__.'/temp/large_header.pdf';
$lorem = file_get_contents('http://loripsum.net/api/html?p=150');
Browsershot::html('<h1>Hello, World!</h1>')
    ->showBrowserHeaderAndFooter()
    ->headerHtml($lorem)
    ->save($targetPath);
```
However, even with a header spanning 150 paragraphs of lorem ipsum text, this finishes in under two seconds, and creates the PDF just fine.

Judging by the error you included, what's missing is the command-line argument including the JSON holding the data that gets passed to the browser instance. It looks like it's there, but for some reason, the keys and values (all strings) have two sets of double quotes (") around them. i.e `{""url"":""path""}`. Maybe this is your issue, but I don't know how it could arise.

When providing URLs, you should also not provide URLs to files, as indicated by this snippet from the `Browsershot` class:
```php
if (str_starts_with(strtolower($url), 'file://')) {
    throw FileUrlNotAllowed::make();
}
```
I believe the JSON issue is preventing this exception from throwing, as the string actually begins with `"file://`.

**Reply by andrecolza** [reply]
@bluesheep100 Unfortunatly the lorem ipsum API only generates 3 short paragraphs, not 150.

Try this:
```php
$targetPath = __DIR__.'/temp/large_header.pdf';
$randomString = Str::random(8000); // using Laravel helper, you can use any other method
Browsershot::html('<h1>Hello, World!</h1>')
    ->showBrowserHeaderAndFooter()
    ->headerHtml("<p>$randomString</p>")
    ->save($targetPath);
```
does it still work for you?

I was able to generate PDFs by this method with random strings of abount 7800 characters (always with 7805 chars, sometimes with 7806, never with 7807).

I have no idea why the printed error has two sets of double quotes, but that's not the point, with smaller headers it works, with longer it doesn't.

**Reply by andrecolza** [reply]
For the avoidance of doubt, in any case I am not manually configuring Browsershot to work, but I am using it via the spatie/laravel-pdf package (see this [issue](https://github.com/spatie/laravel-pdf/issues/168)). However, the error is directly attributable to Browsershot as demonstrated by the fact that it also occurs with the code written above.

**Reply by bluesheep100** [reply]
Still works fine for me. I did `bin2hex(random_bytes(4000))` instead, since I don't have the Laravel helper available here.

What version of Puppeteer and Chrome (presumably) are you doing this one?
There's currently some issues with regard to Puppeteer ^23.0.0, which I have an [open PR](https://github.com/spatie/browsershot/pull/870) to fix, so I'm testing this with the next-latest version 22.15.0. Also on Nodejs 20.
Additionally, are you running this in Windows, inside WSL, or Docker with Docker Desktop? If running inside Docker, try using the `noSandbox()` option.

You can try explicitly passing the path to the binaries for Browsershot to use, maybe it's not using the right Chrome version or something.

**Reply by andrecolza** [reply]
Very interesting, I tried using `bin2hex(random_bytes(4000))` in my local environment (Pupperteer 22.15.0, Chrome win64 127.0.6533.119, Windows without WSL or Docker) and as expected it didn't work, but when I tried using WSL it worked fine, thanks!

Anyway the issue doesn't seem related to the Chrome version: I tried to set a wrong binary path and a small header and I got a "binary not found" error, but when I tried with a 8000 characters long header I get the same error reported above, so I think it is thrown before evaluating the Chrome binaries.

Now I will try to generate my complete pdf using WSL, I'll let you know if everything works correctly!
